Framework/CurveFitting/src/Algorithms/PlotPeakByLogValue.cpp

```diff
@@ -226,160 +235,161 @@ void PlotPeakByLogValue::exec() {
 
   double dProg = 1. / static_cast<double>(wsNames.size());
   double Prog = 0.;
+  IFunction_sptr ifun;
+  bool firstInteration = true;
   for (int i = 0; i < static_cast<int>(wsNames.size()); ++i) {
-    InputData data = getWorkspace(wsNames[i]);
+    InputData data = wsNames[i];
 
     if (!data.ws) {
       g_log.warning() << "Cannot access workspace " << wsNames[i].name << '\n';
       continue;
     }
 
-    if (data.i < 0 && data.indx.empty()) {
+    if (data.i < 0) {
       g_log.warning() << "Zero spectra selected for fitting in workspace "
                       << wsNames[i].name << '\n';
       continue;
     }
 
-    int j, jend;
-    if (data.i >= 0) {
-      j = data.i;
-      jend = j + 1;
-    } else { // no need to check data.indx.empty()
-      j = data.indx.front();
-      jend = data.indx.back() + 1;
-    }
-
     if (createFitOutput) {
-      covarianceWorkspaces.reserve(covarianceWorkspaces.size() + jend);
-      fitWorkspaces.reserve(fitWorkspaces.size() + jend);
-      parameterWorkspaces.reserve(parameterWorkspaces.size() + jend);
+      covarianceWorkspaces.reserve(covarianceWorkspaces.size() + data.i + 1);
+      fitWorkspaces.reserve(fitWorkspaces.size() + data.i + 1);
+      parameterWorkspaces.reserve(parameterWorkspaces.size() + data.i + 1);
     }
 
-    dProg /= abs(jend - j);
-    for (; j < jend; ++j) {
-
-      // Find the log value: it is either a log-file value or simply the
-      // workspace number
-      double logValue = 0;
-      if (logName.empty() || logName == "axis-1") {
-        API::Axis *axis = data.ws->getAxis(1);
-        if (dynamic_cast<BinEdgeAxis *>(axis)) {
-          double lowerEdge((*axis)(j));
-          double upperEdge((*axis)(j + 1));
-          logValue = lowerEdge + (upperEdge - lowerEdge) / 2;
-        } else
-          logValue = (*axis)(j);
-      } else if (logName != "SourceName") {
-        Kernel::Property *prop = data.ws->run().getLogData(logName);
-        if (!prop) {
-          throw std::invalid_argument("Log value " + logName +
-                                      " does not exist");
-        }
-        auto *logp = dynamic_cast<TimeSeriesProperty<double> *>(prop);
-        if (!logp) {
-          throw std::runtime_error("Failed to cast " + logName +
-                                   " to TimeSeriesProperty");
+    dProg /= abs(1);
+    if (isMultiDomainFunction) {
```

Comment:
Whilst you've done a good job cleaning this up, this exec needs to be chunked up. It will help you with debugging and readability in the future

E.g

```
for(...){
    if (isMultiDomainFunction){
       runMultiDomain(....)
    } else {
       i fun = inputFunction
}
firstInteration = false;
findFirstLogValue(...);
```
etc.

I would just extract these to methods and not worry about further tidying of the methods (to save your sanity)